From 4ec46cad02673bd27ddafd6c1666bea277133880 Mon Sep 17 00:00:00 2001
From: Andrii Koval <andrii.koval@lge.com>
Date: Wed, 29 Jul 2015 11:15:52 +0300
Subject: [PATCH] Optimize V8's JSON parser

:Release Notes:
Implemented one optimization point to improve memory usage and performance
of JSON parser.

:Detailed Notes:
Main idea for optimization. Add pre-scan string step into SlowScanJsonString()
function to reduce memory allocation for small strings. Before, buffer was
allocated for at least 1024 elements, now allocates buffer lengths as lengths
as input string.

:Testing Performed:

:QA Notes:

:License:
Copyright (c) 2015 LG Electronics, Inc.

:Issues Addressed:
[DRD-6249] - Optimization of SlowScanJsonString(...) functionality in the V8
JSON parser

Change-Id: Iba8531f92a590f7c0db01d2a0e5e05c43acedaba
Reviewed-on: https://gpro.lgsvl.com/115015
Reviewed-by: Andrii Koval <andrii.koval@lge.com>
Tested-by: Andrii Koval <andrii.koval@lge.com>
Reviewed-by: Tigran Avanesov <tigran.avanesov@lge.com>
---
 deps/v8/src/json-parser.cc |  4 +++-
 deps/v8/src/json-parser.h  | 31 +++++++++++++++++++++++++++++++
 2 files changed, 34 insertions(+), 1 deletion(-)

diff --git a/deps/v8/src/json-parser.cc b/deps/v8/src/json-parser.cc
index 7323b71..efba265 100644
--- a/deps/v8/src/json-parser.cc
+++ b/deps/v8/src/json-parser.cc
@@ -723,7 +723,9 @@ Handle<String> JsonParser<seq_one_byte>::SlowScanJsonString(
     Handle<String> prefix, int start, int end) {
   int count = end - start;
   int max_length = count + source_length_ - position_;
-  int length = Min(max_length, Max(kInitialSpecialStringLength, 2 * count));
+  int string_end_position = FindEndOfStr(position_);
+  int itput_str_length = string_end_position - position_ + count;
+  int length = Min(max_length, itput_str_length);
   Handle<StringType> seq_string =
       NewRawString<StringType>(factory(), length, pretenure_);
   // Copy prefix into seq_str.
diff --git a/deps/v8/src/json-parser.h b/deps/v8/src/json-parser.h
index fc60e29..27d33a3 100644
--- a/deps/v8/src/json-parser.h
+++ b/deps/v8/src/json-parser.h
@@ -60,6 +60,37 @@ class JsonParser BASE_EMBEDDED {
 
   INLINE(void Advance());
 
+  // Returns position where current string ends
+  // it doesn't break the parser state (current position and current symbol)
+  inline int FindEndOfStr(int cur_pos) {
+    //store JSON parser state
+    int orig_pos = position_;
+    uc32 orig_c0 = c0_;
+
+    position_ = cur_pos;
+    uc32 prev_c0 = c0_;
+    while (true) {
+      prev_c0 = c0_;
+      Advance();
+
+      if (c0_ == '"' && prev_c0 != '\\') {
+        break;
+      }
+
+      if (c0_ == kEndOfString) {
+        break;
+      }
+    }
+
+    int end_of_str = position_;
+
+    //restore JSON parser state
+    position_ = orig_pos;
+    c0_ = orig_c0;
+
+    return end_of_str;
+  }
+
   // The JSON lexical grammar is specified in the ECMAScript 5 standard,
   // section 15.12.1.1. The only allowed whitespace characters between tokens
   // are tab, carriage-return, newline and space.
